### PR TITLE
Add experimental export() API

### PR DESCRIPTION
**Pull request by gmagogsfm**
PT2 Export API Prototype

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by angelayi** [comment]
Not sure if you and @tugsbayasgalan are coordinated on this but I think he was having some trouble with circular dependencies on the constrain_as_size with dynamo in [https://github.com/pytorch/pytorch/pull/99992](https://l.facebook.com/l.php?u=https%3A%2F%2Fgithub.com%2Fpytorch%2Fpytorch%2Fpull%2F99992&h=AT2-h4wBetdbkzctIkfLwdpCpUoeSdZuZh8tRKxLadwxspBhhrkFX3vHn_26_p81N8GZDvvkjXCGO_XUs4FYnP5LKW4yTnw1QpqbS3Whzv-HEafZtZg7eMVoU6_Gx2ONaHydI0uzp1Y5ot1fUm7ziT8c1Ts)

**Comment by gmagogsfm** [comment]
> Not sure if you and @tugsbayasgalan are coordinated on this but I think he was having some trouble with circular dependencies on the constrain_as_size with dynamo in [https://github.com/pytorch/pytorch/pull/99992](https://l.facebook.com/l.php?u=https%3A%2F%2Fgithub.com%2Fpytorch%2Fpytorch%2Fpull%2F99992&h=AT2-h4wBetdbkzctIkfLwdpCpUoeSdZuZh8tRKxLadwxspBhhrkFX3vHn_26_p81N8GZDvvkjXCGO_XUs4FYnP5LKW4yTnw1QpqbS3Whzv-HEafZtZg7eMVoU6_Gx2ONaHydI0uzp1Y5ot1fUm7ziT8c1Ts)

Yes, he told me about this. This is based on his PR with some changes. I am waiting for him to fix the circular deps issue, land his PR and then rebase, so that I am not blocked to make next steps of changes.

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -98,7 +99,7 @@ def __init__(self, mod: torch.nn.Module, dynamo_ctx):
 
     def _initialize(self):
         # Do this stuff in constructor to lower overhead slightly
-        if isinstance(self._orig_mod.forward, types.MethodType) and skipfiles.check(
+        if isinstance(self._orig_mod.forward, types.MethodType) and check(
```
Comment (fatcat-z) [inline torch/_dynamo/eval_frame.py:102]: Suggest changing the name 'check()' to 'should_skip()' (or something similar) for a better understanding.

File: torch/_export/__init__.py [hunk 2]
```diff
@@ -39,7 +56,276 @@
...
+
+    Args:
+        m: the `nn.Module` or callable to trace.
+
+        args: Tracing example inputs.
+
+        When `m` is an nn.Module, `args` can be
+        1) A dictionary that maps names of method to their tracing example inputs.
+        in this case, all specified methods will be captured.
+        2) A tuple. In this case, `forward` method of `m` will be captured. It is
```
Comment (angelayi) [inline torch/_export/__init__.py:280]:
```suggestion
        2) A tuple. In this case, only the `forward` method of `m` will be captured. It is
```

File: torch/_export/__init__.py [hunk 3]
```diff
@@ -39,7 +56,276 @@
...
+        equivalent to passing {"forward", tuple-type-args}
+
+        When `m` is a non-Module callable, `args` must be a Tuple containing
+        tracing example inputs.
+
+        config: A CaptureConfig object that specifies how to interpret the
+        program being captured.
+
```
Comment (angelayi) [inline torch/_export/__init__.py:288]:
```suggestion
        constraints: A list of constraints on the dynamic arguments specifying 
        their possible range of their shapes
```

File: torch/_export/graph_module.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
+# pyre-strict
```
Comment (angelayi) [inline torch/_export/graph_module.py:1]: Maybe we can get rid of all the pyre stuff?

File: torch/_export/__init__.py [hunk 5]
```diff
@@ -39,7 +56,276 @@
...
+# that exists on the same MultiMethodExportedProgram.
+# TODO(gmagogsfm): Replace ExportedProgram with MultiMethodExportedProgram.
+
+
+@compatibility(is_backward_compatible=False)
+class MultiMethodExportedProgram:
+    def __init__(self, gms: Dict[str, graph_module.ExportGraphModule]):
+        # TODO(gmagogsfm): Support merging use case where user started by creating
+        # an empty MultiMethodExportedProgram and then start adding more
+        # graph modules to it.
```
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply torch/_export/__init__.py:85]: Not quite, right now the MultiMethodExportedProgram still need to have at least one method to begin with, maybe we should allow users manually adding all the methods one by one (Jacob had one use case like that). but it is not a major feature really.

File: torch/_export/graph_module.py (hunk 4 above)
Comment (gmagogsfm) [reply torch/_export/graph_module.py]: Good call, done.

File: torch/_export/__init__.py (hunk 3 above)
Comment (gmagogsfm) [reply torch/_export/__init__.py:289]: Done.

File: torch/_export/graph_module.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+import torch
+import torch.fx as fx
+import torch.fx._pytree as fx_pytree
+import torch.utils._pytree as pytree
+from . import error
+
+ExportGraphModule = fx.GraphModule
+
+
+LeafValue = Union[
```
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply torch/_export/graph_module.py:13]: Done

File: torch/_export/__init__.py [hunk 7]
```diff
@@ -39,7 +57,280 @@
...
+        in this case, all specified methods will be captured.
+        2) A tuple. In this case, only the `forward` method of `m` will be captured.
+        It is equivalent to passing {"forward", tuple-type-args}
+
+        When `m` is a non-Module callable, `args` must be a Tuple containing
+        tracing example inputs.
+
+        config: A CaptureConfig object that specifies how to interpret the
+        program being captured.
+
```
Comment (angelayi) [inline torch/_export/__init__.py:289]: We don't want this right 

```suggestion
```

**Review by angelayi — APPROVED** [review]

**Pull request closed without merging**